### ISSUE TITLE
[Infra]: Migrate Docker base image from python:3.13-slim to python:3.13-alpine

### ISSUE THREAD

**Issue by wlevan3** [wlevan3/ml-platform-engineering-practicum#80]
## 🎯 Objective

Migrate the Docker base image from `python:3.13-slim` (Debian-based) to `python:3.13-alpine` to reduce container size, minimize attack surface, and improve security posture.

**Expected Benefits:**
- 📦 **40-60% smaller image size** (estimated 150-200MB → 80-120MB)
- 🛡️ **Reduced attack surface** (fewer packages, smaller `.trivyignore`)
- 🔒 **Fewer security alerts** (Alpine has ~30 packages vs Debian's ~138)
- ⚡ **Faster CI/CD** (smaller image = faster pulls/pushes)

---

### Current State
- Base image: `python:3.13-slim` (Debian 13 Trixie)
- Multi-stage build: builder + runtime
- `.trivyignore`: 52 CVEs suppressed (25 LOW, 1 MEDIUM)
- Image size: ~200MB (estimated)
- Security alerts: 0 after PR #79

### Why Alpine?
- **Minimal footprint**: Uses musl libc instead of glibc
- **Security-focused**: Designed for containers, smaller attack surface
- **Proven track record**: Industry standard for lightweight containers
- **Active maintenance**: Regular security updates

### Risks & Trade-offs
- ⚠️ **Compatibility**: musl libc vs glibc (99% compatible for our stack)
- ⚠️ **Build time**: May need to compile more packages (NumPy, scikit-learn)
- ⚠️ **Debugging**: Different tooling if issues arise
- ✅ **Mitigation**: Comprehensive testing strategy + rollback plan

---

### Phase 1: Update Dockerfile Base Images
**File**: `Dockerfile`

**Changes Required**:
```dockerfile
# BEFORE:
FROM python:3.13-slim AS builder
FROM python:3.13-slim AS runtime

# AFTER:
FROM python:3.13-alpine AS builder
FROM python:3.13-alpine AS runtime
```

### Phase 2: Install Alpine-Specific Build Dependencies
**Builder stage changes**:

```dockerfile
# BEFORE (Debian):
RUN apt-get update && apt-get install -y --no-install-recommends \
    gcc \
    && rm -rf /var/lib/apt/lists/*

# AFTER (Alpine):
RUN apk add --no-cache \
    gcc \
    musl-dev \
    linux-headers \
    g++
```

**Justification**:
- `musl-dev` - C standard library headers for Alpine
- `linux-headers` - Required for some Python packages
- `g++` - May be needed for scikit-learn compilation

### Phase 3: Update .trivyignore File
**File**: `.trivyignore`

**Expected Changes**:
- Remove Debian-specific CVEs (glibc, systemd, util-linux, apt)
- Add Alpine-specific CVEs if any (musl libc, apk-tools)
- Expected reduction: 52 CVEs → ~10-15 CVEs

**Categories to Remove**:
- CATEGORY 1: glibc (14 alerts)
- CATEGORY 2: util-linux (9 alerts)
- CATEGORY 3: systemd (6 alerts)
- CATEGORY 8: apt (2 alerts)
- Potentially CATEGORY 4: ncurses (if Alpine doesn't include)

### Phase 4: Update CI/CD Pipeline
**File**: `.github/workflows/ci.yml`

**No changes expected** - Trivy scanning should work identically with Alpine base image.

**Verification**: Confirm Docker build step still works:
```yaml
- name: Build Docker image
  run: docker build -t ml-platform-api:latest .
```

---

### 1. Docker Build Tests
- [ ] **Builder stage builds successfully** - No compilation errors
- [ ] **Runtime stage builds successfully** - No missing dependencies
- [ ] **Image size verification** - Confirm 40-60% reduction
- [ ] **Multi-arch support** - Verify linux/amd64 and linux/arm64 (if required)

**Commands**:
```bash
docker build --target builder -t ml-platform-api:alpine-builder .
docker build --target runtime -t ml-platform-api:alpine-runtime .
docker images | grep ml-platform-api  # Compare sizes
```

### 2. Application Functionality Tests
- [ ] **Container starts successfully** - No missing runtime dependencies
- [ ] **Health check passes** - `/health` endpoint returns 200
- [ ] **Model loading works** - No skops.io or joblib issues
- [ ] **Prediction endpoint works** - `/predict` returns correct results
- [ ] **Model info endpoint works** - `/model/info` returns metadata

**Commands**:
```bash
docker run -d -p 8000:8000 --name test-alpine ml-platform-api:alpine-runtime
curl http://localhost:8000/health
curl -X POST http://localhost:8000/predict -H "Content-Type: application/json" \
  -d '{"sepal_length": 5.1, "sepal_width": 3.5, "petal_length": 1.4, "petal_width": 0.2}'
curl http://localhost:8000/model/info
docker stop test-alpine && docker rm test-alpine
```

### 3. Python Package Compatibility Tests
- [ ] **FastAPI imports** - `from fastapi import FastAPI`
- [ ] **scikit-learn works** - Model predictions match expected results
- [ ] **NumPy operations** - No floating-point precision issues
- [ ] **skops.io loading** - Model deserialization works
- [ ] **Pydantic validation** - Request/response schemas work

**Commands**:
```bash
docker run --rm ml-platform-api:alpine-runtime python -c "import fastapi; print(fastapi.__version__)"
docker run --rm ml-platform-api:alpine-runtime python -c "import sklearn; print(sklearn.__version__)"
docker run --rm ml-platform-api:alpine-runtime python -c "import numpy; print(numpy.__version__)"
docker run --rm ml-platform-api:alpine-runtime python -c "import skops.io as sio; print(sio.__version__)"
```

### 4. Security Scanning Tests
- [ ] **Trivy scan (builder)** - Baseline for new vulnerabilities
- [ ] **Trivy scan (runtime)** - Compare against Debian baseline
- [ ] **Verify reduced CVE count** - Confirm fewer alerts than Debian
- [ ] **No new HIGH/CRITICAL alerts** - Ensure migration doesn't introduce regressions

**Commands**:
```bash
trivy image ml-platform-api:alpine-builder --severity HIGH,CRITICAL
trivy image ml-platform-api:alpine-runtime --severity HIGH,CRITICAL
trivy image ml-platform-api:alpine-runtime --format json > trivy-alpine-results.json
```

### 5. Performance Tests
- [ ] **Container startup time** - Compare Alpine vs Debian
- [ ] **Memory usage** - Check RSS memory footprint
- [ ] **Prediction latency** - Ensure no performance regression
- [ ] **Health check response time** - Should be <100ms

**Commands**:
```bash
time docker run --rm ml-platform-api:alpine-runtime python -c "import app.main"
docker stats --no-stream test-alpine  # Memory usage
# Prediction latency: Use Apache Bench or manual timing
time curl http://localhost:8000/predict -X POST -H "Content-Type: application/json" \
  -d '{"sepal_length": 5.1, "sepal_width": 3.5, "petal_length": 1.4, "petal_width": 0.2}'
```

### 6. CI/CD Integration Tests
- [ ] **GitHub Actions workflow passes** - All jobs green
- [ ] **Docker build job succeeds** - No timeout or memory issues
- [ ] **Trivy security scan passes** - No HIGH/CRITICAL blockers
- [ ] **Python tests pass** - No test failures introduced

### 7. Edge Case Tests
- [ ] **Non-root user works** - `appuser` UID 1000 creation succeeds
- [ ] **Virtual environment works** - `/opt/venv` activation succeeds
- [ ] **Volume mounts work** - If models/ is mounted (future use case)
- [ ] **Environment variables work** - `PYTHONUNBUFFERED=1` applied

---

### High-Risk Areas
1. **NumPy/scikit-learn compilation failures**
   - **Risk**: Alpine may require additional build dependencies
   - **Mitigation**: Add `g++`, `gfortran`, `openblas-dev` if needed
   - **Rollback Trigger**: Builder stage fails with compilation errors

2. **musl libc compatibility issues**
   - **Risk**: skops.io or joblib may have glibc-specific code
   - **Mitigation**: Test model loading thoroughly before merging
   - **Rollback Trigger**: Model deserialization fails or predictions incorrect

3. **Performance regressions**
   - **Risk**: musl libc may have different floating-point precision
   - **Mitigation**: Run prediction accuracy tests against known baseline
   - **Rollback Trigger**: Prediction accuracy drops or latency increases >20%

### Medium-Risk Areas
1. **Increased build time**
   - **Risk**: Alpine may need to compile more wheels
   - **Mitigation**: Use `--no-cache-dir` and multi-stage builds
   - **Rollback Trigger**: Build time exceeds 15 minutes (current ~5 min)

2. **New security vulnerabilities**
   - **Risk**: Alpine packages may have different CVEs
   - **Mitigation**: Review Trivy results before merging
   - **Rollback Trigger**: New HIGH/CRITICAL alerts introduced

### Low-Risk Areas
1. **Image size reduction less than expected**
   - **Risk**: May only achieve 30% reduction instead of 50%
   - **Mitigation**: Still a net positive, document actual savings

---

### Trigger Conditions
Rollback to Debian if **any** of these occur:
- Builder stage compilation fails after adding all dependencies
- Model loading/prediction accuracy fails tests
- CI/CD build time exceeds 15 minutes
- New HIGH/CRITICAL security alerts introduced
- Performance regression >20%

### Rollback Steps
1. **Revert Dockerfile changes**:
   ```bash
   git checkout main -- Dockerfile
   ```

2. **Restore .trivyignore** (if modified):
   ```bash
   git checkout main -- .trivyignore
   ```

3. **Rebuild and verify**:
   ```bash
   docker build -t ml-platform-api:latest .
   docker run -p 8000:8000 ml-platform-api:latest
   ```

4. **Document findings** in PR or issue comments for future attempts

---

### Must-Have (Blocking Merge)
- ✅ All 17 tests pass (Docker build, functionality, security, CI/CD)
- ✅ Image size reduced by at least 30%
- ✅ No new HIGH/CRITICAL security alerts
- ✅ Prediction accuracy matches Debian baseline
- ✅ CI/CD pipeline passes all checks

### Nice-to-Have (Non-Blocking)
- 🎯 Image size reduced by 50%+
- 🎯 `.trivyignore` reduced to <20 CVEs
- 🎯 Build time reduced or stays within 10% of current
- 🎯 Container startup time improved

---

## 📅 Implementation Timeline

**Estimated Time**: 4-5 hours

1. **Phase 1: Dockerfile Update** - 30 minutes
   - Update base images
   - Add Alpine build dependencies
   - Test builder and runtime stages

2. **Phase 2: Testing** - 2 hours
   - Run all 17 test categories
   - Document any issues encountered
   - Iterate on build dependencies if needed

3. **Phase 3: Security Analysis** - 1 hour
   - Run Trivy scans
   - Update `.trivyignore` file
   - Document CVE changes

4. **Phase 4: Documentation & PR** - 1 hour
   - Update `CLAUDE.md` references to Alpine
   - Create PR with comprehensive description
   - Run CI/CD and verify all checks pass

5. **Phase 5: Review & Merge** - 30 minutes
   - Address any PR feedback
   - Merge to main once approved

---

### Documentation
- [Alpine Linux Packages](https://pkgs.alpinelinux.org/packages)
- [Python Alpine Docker Images](https://hub.docker.com/_/python)
- [musl libc Documentation](https://musl.libc.org/)
- [Alpine Linux Best Practices](https://wiki.alpinelinux.org/wiki/Best_Practices)

### Relevant Files
- `Dockerfile` - Multi-stage build definition
- `.trivyignore` - Security vulnerability suppressions
- `requirements.txt` - Python dependencies
- `.github/workflows/ci.yml` - CI/CD pipeline
- `docs/VULNERABILITY_REMEDIATION.md` - Security workflow

### Related Issues/PRs
- PR #79 - CVE-2025-7709 suppression (100% alert remediation)
- Issue #12 - Docker security scanning implementation

---

## 📝 Notes

- This migration is part of **Phase 3** of the project roadmap
- Current `.trivyignore` has 52 CVEs (51 for Debian packages, 1 for pip - already fixed)
- Alpine migration aligns with project goal of production-grade ML platform
- Consider Distroless as alternative in future if Alpine has issues
- Document all findings and learnings in PR for future reference

**Comment by wlevan3** [comment]
# Alpine Migration Decision - Issue #80

**Date**: 2025-11-03
**Decision**: **DO NOT PROCEED** with Alpine migration  
**Recommendation**: Close issue with detailed findings and learnings documented

---

## Executive Summary

The Alpine migration investigation revealed that **Alpine is NOT the optimal choice** for scientific Python workloads despite initial assumptions. While technically functional and offering improved security, the tradeoffs significantly undermine the stated goals.

**Key Findings:**
- ✅ **Security Winner**: Alpine eliminates 66.67% of vulnerabilities (3 → 1)
- ❌ **Size Disappointment**: Only 22% reduction vs 40-60% target (758MB → 590MB = 168MB saved)
- ❌ **Build Time Penalty**: 2x slower builds (~5 min → ~10 min)
- ❌ **Maintenance Complexity**: Requires Alpine-specific runtime dependencies

**Verdict**: The security benefit does NOT justify the size/build time penalties for a scientific Python stack.

---

## Detailed Comparison

### Security Analysis

| Metric | Debian (python:3.13-slim) | Alpine (python:3.13-alpine) | Improvement |
|--------|---------------------------|------------------------------|-------------|
| **Total Vulnerabilities** | 3 (MEDIUM: 2, HIGH: 1) | 1 (MEDIUM: 1) | **-66.67%** ✅ |
| **CRITICAL** | 0 | 0 | No change |
| **HIGH** | 1 | 0 | **-100%** ✅ |
| **MEDIUM** | 2 | 1 | **-50%** ✅ |
| **LOW** | 0 (suppressed via `.trivyignore`) | 0 | No change |

**Debian Vulnerabilities:**
1. CVE-2025-8869 (pip 25.2 → 25.3) - MEDIUM - **Fixed** in both images
2. 2 additional vulnerabilities (details in Trivy scan)

**Alpine Vulnerabilities:**
1. CVE-2025-8869 (pip 25.2 → 25.3) - MEDIUM - **Fixed** in Alpine image

**Security Conclusion**: Alpine reduces vulnerabilities by 66.67%, eliminating all HIGH severity issues. This is a significant security improvement.

---

### Image Size Analysis

| Metric | Debian | Alpine | Change |
|--------|--------|--------|--------|
| **Size** | 758MB | 590MB | **-168MB (-22%)** ❌ |
| **Expected** | 758MB | 303-455MB | **-303-455MB (-40-60%)** |
| **Gap** | - | **-135-287MB shortfall** | **18-38% below target** |

**Why the Gap?**
- **Pre-built wheels on Debian**: NumPy, SciPy, scikit-learn ship as optimized binary wheels
- **Source compilation on Alpine**: All scientific packages must compile from source
- **Binary bloat**: Compiled binaries from source are larger than pre-built wheels
- **musl libc overhead**: Alpine's musl libc creates larger binaries than glibc for scientific code

**Size Conclusion**: Alpine's 22% reduction is **significantly below** the 40-60% target, making it a poor choice for size optimization.

---

### Build Time Analysis

| Metric | Debian | Alpine | Change |
|--------|--------|--------|--------|
| **Build Time** | ~5 minutes | ~10+ minutes | **+100% (2x slower)** ❌ |
| **scikit-learn compile** | N/A (pre-built) | 172+ seconds | New bottleneck |
| **NumPy compile** | N/A (pre-built) | ~60 seconds | New bottleneck |
| **SciPy compile** | N/A (pre-built) | ~45 seconds | New bottleneck |

**Why 2x Slower?**
- **Wheel availability**: Debian uses PyPI's pre-built wheels; Alpine must compile from source
- **Compilation overhead**: gcc + g++ + build dependencies add significant time
- **CI cost**: 2x build time = 2x GitHub Actions minutes consumed per PR

**Build Time Conclusion**: Alpine's 2x build time penalty is **unacceptable** for CI/CD workflows, slowing down development velocity.

---

### Functional Testing Results

| Test | Debian | Alpine | Status |
|------|--------|--------|--------|
| Container starts | ✅ | ✅ | **PASS** |
| Health check (`/health`) | ✅ | ✅ | **PASS** |
| Model loading (skops.io) | ✅ | ✅ | **PASS** |
| Prediction (`/predict`) | ✅ | ✅ | **PASS** |
| FastAPI import | ✅ | ✅ | **PASS** |
| scikit-learn import | ✅ | ✅ | **PASS** |
| NumPy import | ✅ | ✅ | **PASS** |
| skops.io import | ✅ | ✅ | **PASS** |

**Functional Conclusion**: Alpine is **100% functionally compatible**. No regressions detected.

---

### Complexity & Maintenance

**[View full analysis document]** (4000+ characters): /tmp/alpine-migration-decision.md

**TL;DR**: Alpine migration **NOT RECOMMENDED** for this ML platform due to:
- ❌ Only 22% size reduction (vs 40-60% target) - **shortfall of 18-38%**
- ❌ 2x slower builds (~5 min → ~10 min) - unacceptable for CI/CD velocity
- ❌ Alpine-specific complexity (package manager, runtime deps, user management)
- ✅ 66.67% CVE reduction (3 → 1) - **good but not worth the tradeoffs**

**Verdict**: **Debian + `.trivyignore`** wins on weighted decision matrix (7.55/10 vs 7.35/10).

**Key Learning**: For scientific Python workloads, **application dependencies dominate image size**, not the base OS. Alpine's smaller OS footprint is overshadowed by large compiled binaries from source.

**Recommendation**: Continue using Debian; defer Distroless exploration to Phase 3+.
